### ISSUE TITLE
Istio Upgrades Cause Spike in Istio-Proxy Memory

### ISSUE THREAD

**Issue by crhuber** [istio/istio#23959]
**Bug description**

We have a large istio deployment with ~1000 connected istio-proxy's. 

When upgrading from Istio 1.4.7 > 1.5.4 (or any other version upgrade) we experience a large increase of memory use per istio-proxy for a few hours after the deployment. The net result is that the memory use spike on roughly 1000 istio-proxy's causes Kubernetes nodes to be starved of memory and eventually go to `NotReady` state. Also during this time Pilot cannot push to istio-proxy's  and we see frequent

`RDS: Send failure for node:` or `EDS: Send failure for node:`

presumably because the Kubernetes node is starved for memory and istio-proxy cannot respond. This then results in most istio-proxy's going to `STALE` when running `istioctl proxy-status`. 
Then our services start to fail with frequent 503 errors

5xx Errors:

![image](https://user-images.githubusercontent.com/5857391/82350552-cb9b1200-99fb-11ea-97b3-0de3011c13cf.png)

As seen below during deployment at `15:10`, istio-proxy memory shoots from 189Mi to 291Mi. Then gradually over the course of more than >4 hours the memory use goes back to lower number.

However, this doesnt seem to be limited to 1.4.7 or 1.5.4 because we also had the same problem in 1.4.3 when enabling auto mTLS and raised an issue here: https://github.com/istio/istio/issues/20797. (We mitigated that issue by giving each node more free memory)

**istio-proxy memory**
Memory jumps from 189Mi to 291Mi per proxy
![image](https://user-images.githubusercontent.com/5857391/82349663-afe33c00-99fa-11ea-9b56-1178e4f67c46.png)

**istio-proxy cpu**
Also we have a spike in CPU use
![image](https://user-images.githubusercontent.com/5857391/82349976-09e40180-99fb-11ea-83c9-a54009903143.png)

**Pilot Pushes**
A large spike in Pilot Pushes
![image](https://user-images.githubusercontent.com/5857391/82350659-eff6ee80-99fb-11ea-9a7e-af02a60ec069.png)

**xDS request size**
XDS Request Max Bytes spikes at that time 
(I'm only collecting envoy metrics for about 30 sidecars - but for the few that we are scraping this is the metric i saw)

![image](https://user-images.githubusercontent.com/5857391/82350762-0d2bbd00-99fc-11ea-8e8c-749fbcb2b903.png)

**Pilot CPU & Memory**
High amount of CPU and Memory being used on Pilot
![image](https://user-images.githubusercontent.com/5857391/82351123-74e20800-99fc-11ea-833e-e8d155880518.png)

How can we safely deploy upgrades when doing so effectively takes down our Kube cluster. Are there any suggestions on how to prevent this?

We have:

* 1042 services (not all are in the mesh)
* 1155 istio-proxy
* 95 nodes (4 VCPU, 8 GB)
* 13-18 avg istio-proxy per node

**Expected behavior**
Upgrades Do Not Cause Memory Spike

**Steps to reproduce the bug**
NA

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

```
citadel version: 1.5.4
galley version: 1.5.4
ingressgateway version: 1.5.4
pilot version: 1.5.4
sidecar-injector version: 1.5.4
telemetry version: 1.5.4
data plane version: 1.4.7 (401 proxies), 1.5.4 (754 proxies)
```

**How was Istio installed?**
helm

**Environment where bug was observed (cloud vendor, OS, etc)**
Amazon EKS .
Kube 1.14

**Comment by juniorz** [comment]
> istio-proxy memory
> Memory jumps from 189Mi to 291Mi per proxy

This graph is for memory use across all sidecars. When we aggregate by image version, we see that `1.4.7` proxies had a ~`100M` increase after the control plane was upgraded, and the `1.5.4` sidecars had only a slight increase in memory use. 

![image](https://user-images.githubusercontent.com/30809/82461848-23e51900-9abb-11ea-8927-6e23c90a46fc.png)

Our current hypothesis is that the `1.4.7` proxies are not releasing resources (config, maybe) related to the previous control plane version.

**Comment by sdake** [comment]
@crhuber first off, thank you for adopting Istio into such a large deployment!

Are the Sidecar objects in use?

I have copied @mandarjog and @douglas-reid as the two foremost experts in the project around memory consumption. Will get others involved as needed.

Cheers
-steve

**Comment by mandarjog** [comment]
@crhuber thanks for the analysis.
1. How many namespaces do you have?
2. 200MB Steady state mem for proxy is higher than usual for a sidecar.
3. How are you upgrading ? Rolling restart one deployment at a time or more ?
4. Can you get a config dump from the older proxy 
5. Can you also get a head dump? https://github.com/istio/istio/wiki/Analyzing-Istio-Performance#memory-profile

**Comment by crhuber** [comment]
@mandarjog 

0. Yes Sidecar objects in use
1. We have 34 namespaces. 14 namespace are `istio-injection=enabled`
2. I agree
3. To upgrade we are using the helm chart. We just bump the chart version (ie: to 1.5.4) and Kubernetes proceedes with the upgrade using the default deployment upgrade parameters in the chart
4. I sent config dump of 2 older proxies and 1 pilot
5. Working on it
6. Working on it

**Comment by crhuber** [comment]
@mandarjog any update on this one?

**Comment by howardjohn** [comment]
> Our current hypothesis is that the 1.4.7 proxies are not releasing resources (config, maybe) related to the previous control plane version.

This is correct. It was added in version 1.6. Have you had a chance to try this out in 1.6+?

**Comment by howardjohn** [comment]
This is presumed fixed in 1.6 per above comment. If not, we can re-open and explore further.
